tools/codegen/gen.py

```diff
@@ -312,31 +312,65 @@ def go(f: NativeFunction) -> Optional[str]:
+                result += f"CAFFE2_API {cpp_returns_type} {name}({cpp_binding.signature.no_defaults().cpp_arguments_str()});"
+
+            return result
 
         assert target is Target.DEFINITION
 
-        dispatcher_exprs = dispatcher.cpparguments_exprs(cpp_args)
-        cpp_args_str_no_default = ', '.join(map(lambda a: a.str_no_default(), cpp_args))
         dispatcher_returns_type = dispatcher.returns_type(f.func.returns)
-        dispatcher_types_str = ', '.join(map(lambda a: a.type, dispatcher_exprs))
-        dispatcher_exprs_str = ', '.join(map(lambda a: a.expr, dispatcher_exprs))
 
-        return f"""
+        if cpp_binding.gathered_signature is None:
+            # There's no TensorOptions
+            return f"""
 // aten::{f.func}
-{cpp_returns_type} {name}({cpp_args_str_no_default}) {{
+{cpp_returns_type} {name}({cpp_binding.signature.no_defaults().cpp_arguments_str()}) {{
     static auto op = c10::Dispatcher::singleton()
         .findSchemaOrThrow("aten::{f.func.name.name}", "{f.func.name.overload_name}")
-        .typed<{dispatcher_returns_type} ({dispatcher_types_str})>();
-    return op.call({dispatcher_exprs_str});
+        .typed<{dispatcher_returns_type} ({cpp_binding.signature.types_str()})>();
+    return op.call({cpp_binding.signature.exprs_str()});
+}}
+"""
+        elif local.use_c10_dispatcher() is UseC10Dispatcher.full:
+            # for c10-full ops, the scattered version is the real op and the gathered version is a proxy
+            # calling into the scattered version
+            return f"""
+// aten::{f.func}
+{cpp_returns_type} {name}({cpp_binding.signature.no_defaults().cpp_arguments_str()}) {{
+    static auto op = c10::Dispatcher::singleton()
+        .findSchemaOrThrow("aten::{f.func.name.name}", "{f.func.name.overload_name}")
+        .typed<{dispatcher_returns_type} ({cpp_binding.signature.types_str()})>();
+    return op.call({cpp_binding.signature.exprs_str()});
+}}
+{cpp_returns_type} {name}({cpp_binding.gathered_signature.no_defaults().cpp_arguments_str()}) {{
+    return {name}({cpp_binding.gathered_signature.exprs_str(dispatcher.ProcessTensoroptions.SCATTER)});
 }}
 """
+        else:
+            # for non-c10-full ops, the gathered version is the real op and the scattered version is a proxy
+            # calling into the gathered version
+            return f"""
+// aten::{f.func}
+{cpp_returns_type} {name}({cpp_binding.gathered_signature.no_defaults().cpp_arguments_str()}) {{
+    static auto op = c10::Dispatcher::singleton()
+        .findSchemaOrThrow("aten::{f.func.name.name}", "{f.func.name.overload_name}")
+        .typed<{dispatcher_returns_type} ({cpp_binding.gathered_signature.types_str()})>();
+    return op.call({cpp_binding.gathered_signature.exprs_str()});
+}}
+{cpp_returns_type} {name}({cpp_binding.signature.no_defaults().cpp_arguments_str()}) {{
+    return {name}({cpp_binding.gathered_signature.exprs_str(dispatcher.ProcessTensoroptions.GATHER)});
```

Comment:
I made `exclude_this` and `with_defaults` into parameters instead of functions. I know this only partially addresses your issue but it's the best I came up with after several attempts.